dds/DCPS/XTypes/TypeObject.cpp

```diff
@@ -272,6 +272,119 @@ TypeIdentifier makeTypeIdentifier(const TypeObject& type_object, const DCPS::Enc
   return ti;
 }
 
+AnnotationParameterValue::AnnotationParameterValue(ACE_CDR::Octet kind)
+  : kind_(kind)
+  , active_(0)
+{
+  activate();
+}
+
+AnnotationParameterValue::AnnotationParameterValue(const AnnotationParameterValue& other)
+  : kind_(other.kind_)
+  , active_(0)
+{
+  activate(&other);
+}
+
+AnnotationParameterValue& AnnotationParameterValue::operator=(const AnnotationParameterValue& other)
+{
+  if (&other == this) {
+    return *this;
+  }
+  reset();
+  kind_ = other.kind_;
+  activate(&other);
+  return *this;
+}
+
+void AnnotationParameterValue::activate(const AnnotationParameterValue* other)
+{
+#define OPENDDS_BRANCH_ACTIVATE(T, N) \
+  active_ = new(N ## _) T;            \
+  if (other) N() = other->N();        \
+  break
+
+  switch (kind_) {
+  case TK_NONE:
+    break; // no-op, no member selected
+  case TK_BOOLEAN:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Boolean, boolean_value);
+  case TK_BYTE:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Octet, byte_value);
+  case TK_INT16:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Short, int16_value);
+  case TK_INT32:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Long, int32_value);
+  case TK_INT64:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::LongLong, int64_value);
+  case TK_UINT16:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::UShort, uint16_value);
+  case TK_UINT32:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::ULong, uint32_value);
+  case TK_UINT64:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::ULongLong, uint64_value);
+  case TK_FLOAT32:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Float, float32_value);
+  case TK_FLOAT64:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Double, float64_value);
+  case TK_FLOAT128:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::LongDouble, float128_value);
+  case TK_INT8:
+  case TK_UINT8:
+    break; // no-op, no member selected
+  case TK_CHAR8:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Char, char_value);
+  case TK_CHAR16:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::WChar, wchar_value);
+  case TK_ENUM:
+    OPENDDS_BRANCH_ACTIVATE(ACE_CDR::Long, enumerated_value);
+  case TK_STRING8:
+    OPENDDS_BRANCH_ACTIVATE(OPENDDS_STRING, string8_value);
+  case TK_STRING16:
+    OPENDDS_BRANCH_ACTIVATE(OPENDDS_WSTRING, string16_value);
+  default:
+    OPENDDS_BRANCH_ACTIVATE(ExtendedAnnotationParameterValue, extended_value);
+  }
+}
+
+void AnnotationParameterValue::reset()
+{
+  using DCPS::String;
+  using DCPS::WString;
+
+  if (!active_) {
+    return;
+  }
+  #define OPENDDS_BRANCH_RESET(T) static_cast<T*>(active_)->~T(); break
+  switch (kind_) {
+  case TK_NONE:
+    break; // no-op, no member selected
+  case TK_BOOLEAN:
+  case TK_BYTE:
+  case TK_INT16:
+  case TK_INT32:
+  case TK_INT64:
+  case TK_UINT16:
+  case TK_UINT32:
+  case TK_UINT64:
+  case TK_FLOAT32:
+  case TK_FLOAT64:
+  case TK_FLOAT128:
+  case TK_INT8:
+  case TK_UINT8:
+  case TK_CHAR8:
+  case TK_CHAR16:
+  case TK_ENUM:
+    break; // no-op, no member selected
```

Comment:
These members are primitive and don't have a destructor.

Reply:
It's a trivially-destructible type so we don't have a way to do the `ptr->~Type()` expression, nor do we need it.

However, this is a different usage than the "no member selected" case that happens in TypeObject.  

TypeObject is an IDL union that has certain values of its discriminator which are meaningful but don't "select a member."  (The spec should use something like an empty struct, which is allowed by the "extended" building block, for this instead of making it a comment...)

Here in AnnotationParameterValue we have union branches that do "select a member" and that member is of a basic type.  So instead they should use the array storage to make an object of that type like `int* p = new(storage) int();` which formally begins the lifetime of the object and value-initializes it to 0.  The compiler will probably optimize it out, but it may help Coverity/valgrind/asan/etc.

Also, while researching it on cppreference I found that the storage array should really be `unsigned char` instead of plain `char`.